### ISSUE TITLE
Question about loss landscape

### ISSUE THREAD

**Issue by keirawu011029** [huanranchen/AdversarialAttacks#5]

**Comment by keirawu011029** [comment]
Dear author:
I ues the original code(experiments/attacks/DrawLandscape.py) to plot the loss landscape on target model, but I didn't get the same figure that you showed in the paper. I got:
![image](https://github.com/user-attachments/assets/27d65052-6bb0-4b72-83bc-3b29a24116e2)

I wanna know how I can fix the problem

**Comment by huanranchen** [comment]
> Dear author: I ues the original code(experiments/attacks/DrawLandscape.py) to plot the loss landscape on target model, but I didn't get the same figure that you showed in the paper. I got: ![image](https://private-user-images.githubusercontent.com/80230620/351224202-27d65052-6bb0-4b72-83bc-3b29a24116e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE3NDU0OTYsIm5iZiI6MTcyMTc0NTE5NiwicGF0aCI6Ii84MDIzMDYyMC8zNTEyMjQyMDItMjdkNjUwNTItNmJiMC00YjcyLTgzYmMtM2IyOWEyNDExNmUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIzVDE0MzMxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhMTlkMDY4ZjcwMTg1ZTI2OTk5YmQ2NDY4OTcwZDYxNzcyYTk1ZDFjNzg4ZTg1NWU0ZmI1MTE4YThkMTExYTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.E_ovHjYOXQzXkUUIzCm-Qjvdi2SMboELccZhURjfVWw)
> 
> I wanna know how I can fix the problem

@keirawu011029 

Hi! Thank you for your interest! You only need to change 
```
direction = (now_x - x)
```
to
```
direction = (now_x - x) * (-1) ** i
```
.

I have also committed this change, you can use git fetch to get the newest code.


The issue comes from the fact that we use the worst-case direction (now_x-x) to draw the landscape, rather than the average-case direction. In the positive direction x + a (now_x-x) with a>0, the loss will become much larger. In the negative direction x + a (now_x-x) with a<0 the loss will become smaller since it's adding adversarial perturbations to the input x.

So in a nutshell, since we use the worst-case direction to draw the landscape (now_x-x), the loss landscape is not symmetric. You can flip some landscape (i.e., using direction*= (-1)**i) to make it look more symmetric. Or, just draw the landscape in average direction, or two (nearly) worst case directions to make it symmetric.
